src/mpid/ch4/generic/gpu/mpidig_gpu_impl.h

```diff
@@ -0,0 +1,173 @@
+    size_t size;
+    MPI_Aint true_lb;
+} MPIDIG_gpu_stage_ctx_t;
+
+/* MPIDIG_gpu_stage_send_buf - allocate stage buffer, copy source data into it, and return pointer */
+MPL_STATIC_INLINE_PREFIX void *MPIDIG_gpu_stage_send_buf(const void *buf, MPI_Aint count,
+                                                         MPI_Datatype datatype)
+{
+    char *tmp_buf = (char *) buf;
+#ifdef HAVE_GPU_STAGING
+    int dt_contig;
+    size_t size;
+    MPIR_Datatype *dt_ptr;
+    MPI_Aint true_lb;
+
+    MPIDI_Datatype_get_info(count, datatype, dt_contig, size, dt_ptr, true_lb);
+    if (!dt_contig)
+        size = (dt_ptr->extent * count) + (dt_ptr->true_ub - true_lb) - dt_ptr->extent;
+
+    tmp_buf = MPIDIG_gpu_get_buf(size);
+    MPL_gpu_memcpy_d2h(tmp_buf, (char *) buf + true_lb, size);
+    tmp_buf -= true_lb;
+
+    MPIDIG_gpu_stage_ctx_t *ctx = MPL_malloc(sizeof(*ctx), MPL_MEM_OTHER);
+    MPIR_Assert(ctx != NULL);
+    ctx->true_lb = true_lb;
+    MPIDIU_map_set(MPIDIG_gpu_global.buf_map, (uint64_t) tmp_buf, ctx, MPL_MEM_OTHER);
+#endif
+    return tmp_buf;
+}
+
+/* MPIDIG_gpu_unstage_send_buf - free stage buffer */
+MPL_STATIC_INLINE_PREFIX void MPIDIG_gpu_unstage_send_buf(const void *buf)
+{
+#ifdef HAVE_GPU_STAGING
+    MPIDIG_gpu_stage_ctx_t *ctx =
+        (MPIDIG_gpu_stage_ctx_t *) MPIDIU_map_lookup(MPIDIG_gpu_global.buf_map, (uint64_t) buf);
+    MPIDIG_gpu_release_buf((char *) buf + ctx->true_lb);
+    MPL_free(ctx);
+    MPIDIU_map_erase(MPIDIG_gpu_global.buf_map, (uint64_t) buf);
+#endif
+}
+
+/* MPIDIG_gpu_stage_recv_buf - allocate stage buffer and return pointer */
+MPL_STATIC_INLINE_PREFIX void *MPIDIG_gpu_stage_recv_buf(void *buf, MPI_Aint count,
+                                                         MPI_Datatype datatype)
+{
+    char *tmp_buf = (char *) buf;
+#ifdef HAVE_GPU_STAGING
+    int dt_contig;
+    size_t size;
+    MPIR_Datatype *dt_ptr;
+    MPI_Aint true_lb;
+
+    MPIDI_Datatype_get_info(count, datatype, dt_contig, size, dt_ptr, true_lb);
+    if (!dt_contig)
+        size = (dt_ptr->extent * count) + (dt_ptr->true_ub - true_lb) - dt_ptr->extent;
+
+    tmp_buf = MPIDIG_gpu_get_buf(size);
+    tmp_buf -= true_lb;
+
+    MPIDIG_gpu_stage_ctx_t *ctx = MPL_malloc(sizeof(*ctx), MPL_MEM_OTHER);
+    MPIR_Assert(ctx != NULL);
+    ctx->src_buf = buf;
+    ctx->size = size;
+    ctx->true_lb = true_lb;
+    MPIDIU_map_set(MPIDIG_gpu_global.buf_map, (uint64_t) tmp_buf, ctx, MPL_MEM_OTHER);
+#endif
+    return tmp_buf;
+}
+
+/* MPIDIG_gpu_unstage_recv_buf - copy data from stage buffer to source buffer and free it */
+MPL_STATIC_INLINE_PREFIX void MPIDIG_gpu_unstage_recv_buf(void *buf)
+{
+#ifdef HAVE_GPU_STAGING
+    MPIDIG_gpu_stage_ctx_t *ctx =
+        (MPIDIG_gpu_stage_ctx_t *) MPIDIU_map_lookup(MPIDIG_gpu_global.buf_map, (uint64_t) buf);
+    void *stage_buf = (char *) buf + ctx->true_lb;
+    void *src_buf = (char *) ctx->src_buf + ctx->true_lb;
+    MPL_gpu_memcpy_h2d(src_buf, stage_buf, ctx->size);
```

Comment:
`MPL_gpu_memcpy_h2d` will be replaced by `yaksa_ipack` (or some other `MPIR_xxx` wrapper interface for yaksa).